### DISCUSSION TITLE
Metrics Update Frequency

### DISCUSSION THREAD

**Discussion by marvin-steinke** [sustainable-computing-io/kepler#1574] [Q&A] [answered]
How often are metrics like RAPL and cpu usage updated/measured and is it possible to configure this frequency?

**Comment by dave-tucker** [comment, marked as answer]
Every 3 seconds.
Yes you can pass the following environment variable: `SAMPLE_PERIOD_SEC=5" to change this to 5 seconds.

**Reply by marvin-steinke** [reply]
Is 1 the lowest value for this variable?

**Reply by dave-tucker** [reply]
Yes

**Reply by marvin-steinke** [reply]
How does the measurements internally work with the hardware counters? Is there a counter for the total energy consumed and then you measure this at two points which are subtracted to get the amount of energy consumed between these two points; i.e. do two experiments with a sampling period of 1s and 30s respectively added up over a 60s period result in the same total energy consumption?

**Reply by dave-tucker** [reply]
The counters are monotonically increasing.
`SAMPLE_PERIOD_SEC` dictates how often they are updated or how fine-grained the counters are.

In other words
- SAMPLE_PERIOD_SEC=1 will update the counters every second
- SAMPLE_PERIOD_SEC=60 will update the counters every minute

If you view these counters after 10s, the former has been updated 10 times where the latter has updated 0 times.
After 70 seconds, the former has been updated 70 times, where the latter has only updated 1 time.

**Reply by marvin-steinke** [reply]
But will both counters have the same value at 60s runtime?

**Reply by dave-tucker** [reply]
Yes!
